x-pack/plugins/actions/server/create_execute_function.ts

```diff
@@ -119,6 +132,92 @@ export function createExecutionEnqueuerFunction({
   };
 }
 
+export function createBulkExecutionEnqueuerFunction({
+  taskManager,
+  actionTypeRegistry,
+  isESOCanEncrypt,
+  preconfiguredActions,
+}: CreateExecuteFunctionOptions): BulkExecutionEnqueuer<void> {
+  return async function execute(
+    unsecuredSavedObjectsClient: SavedObjectsClientContract,
+    options: ExecuteOptions[]
+  ) {
+    if (!isESOCanEncrypt) {
+      throw new Error(
+        `Unable to execute action because the Encrypted Saved Objects plugin is missing encryption key. Please set xpack.encryptedSavedObjects.encryptionKey in the kibana.yml or use the bin/kibana-encryption-keys command.`
+      );
+    }
```

Comment:
Often when we're executing actions, we're using the same connector for multiple alerts. For example if a rule has 1 action attached and it generates 500 alerts, we're using that same connector 500 times. It might be a nice optimization to split some of this processing so we're only doing `getAction` (which actually means `getConnector` once for each connector id instead of for each action.

Could do something like this:

```
const connectorIds = [...new Set(options.map((action) => action.id))];
const connectors = await getConnectors(
      unsecuredSavedObjectsClient,
      preconfiguredActions,
      connectorIds
 );

// validate access to connector (validateCanActionBeUsed)
// validate access to connector type (actionTypeRegistry.isActionExecutable)

```

where `getConnectors` returns all the unique connectors used using the saved object bulk get API
```
async function getConnectors(
  unsecuredSavedObjectsClient: SavedObjectsClientContract,
  preconfiguredConnectors: PreConfiguredAction[],
  connectorIds: string[]
): Promise<GetConnectorsResult[]> {
  const result: GetConnectorsResult[] = [];

  const connectorIdsToFetch = [];
  for (const connectorId of connectorIds) {
    const pcConnector = preconfiguredConnectors.find((connector) => connector.id === connectorId);
    if (pcConnector) {
      result.push({ connector: pcConnector, isPreconfigured: true, id: connectorId });
    } else {
      connectorIdsToFetch.push(connectorId);
    }
  }

  if (connectorIdsToFetch.length > 0) {
    const bulkGetResult = await unsecuredSavedObjectsClient.bulkGet<RawAction>(
      connectorIdsToFetch.map((id) => ({
        id,
        type: 'action',
      }))
    );

    for (const item of bulkGetResult.saved_objects) {
      if (item.error) throw item.error;
      result.push({
        isPreconfigured: false,
        connector: item.attributes,
        id: item.id,
      });
    }
  }

  return result;
}
```

and then after all the connector processing is done, then we iterate over the actions to do all the other stuff.

LMK if that makes sense! Happy to walk through it.

Reply:
Yes that does, thank you! I will update this 🙂

Reply:
Resolved in this [commit](https://github.com/elastic/kibana/pull/137781/commits/b15ed37052ae2e0e865df502af4255665c206408)